test/test.downscopedclient.ts

```diff
@@ -0,0 +1,406 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as assert from 'assert';
+import {describe, it, before, after, beforeEach, afterEach} from 'mocha';
+import * as nock from 'nock';
+import * as sinon from 'sinon';
+
+import {StsSuccessfulResponse} from '../src/auth/stscredentials';
+import {DownscopedClient} from '../src/auth/downscopedclient';
+import {AuthClient} from '../src/auth/authclient';
+import {mockStsBetaTokenExchange} from './externalclienthelper';
+import {GoogleAuth} from '../src';
+
+nock.disableNetConnect();
+
+interface SampleResponse {
+  foo: string;
+  bar: number;
+}
+
+describe('DownscopedClient', () => {
+  let clock: sinon.SinonFakeTimers;
+
+  const auth = new GoogleAuth({
+    keyFilename: './test/fixtures/private.json',
+    scopes: 'https://www.googleapis.com/auth/cloud-platform',
+  });
+  let client: AuthClient;
```

Comment:
I think you don't need to do this. This reads from a real location and generates tokens.
It is better to define a dummy class which meets the AuthClient interface.

See how I did it in Python: https://github.com/googleapis/google-auth-library-python/blob/63ac08afe44c3da51a6fb3bfc2fcdb6588e040ef/tests/test_downscoped.py#L82 

This will also simplify the implementation below. You wouldn't need to stub its underlying implementation if you do so.

Reply:
Yea, I understand and have thought about this. 
In this case, language might differ, this approach could save a 3 line stub for calling getAccessToken().
However, AuthClient has 3 abstract APIs, implementing the other two abstracts APIs getRequestHeader and request which won't get called in tests is less efficient compared to calling stubs.